### ISSUE TITLE
ImGui Plotting Glitch with modified plot function (overflowing default 64k vertices mesh limit)

### ISSUE THREAD

**Issue by salmanjalali** [ocornut/imgui#1727]
Currently, I have a window with several plot widgets going all at once. Though, after increasing the buffer size up to a threshold, the GUI glitches and is not able to render at that refresh rate. 

I am able to increase this threshold by decreasing the refresh rate more and more, but I do not see this as an answer. I was wondering if this is known issue in the GL pipeline, or if there is any flag or anything that needs to be set to mitigate this issue.

The glitch looks as follows:
![broken-gui](https://user-images.githubusercontent.com/10891652/38449540-237e2fb4-39c5-11e8-9742-4216aa8a1fe1.png)

The code used to plot looks pretty regular:
```
    if (m_plotDuration < 30) {
        m_windowRefreshRate = DEFAULT_REFRESH_RATE;
    } else {
        m_windowRefreshRate = DEFAULT_REFRESH_RATE - ((m_plotDuration - 30) % 10);
    }
    size_t newVecSize = static_cast<size_t>(m_windowRefreshRate * m_plotDuration);
    m_ioSpeedVec.resize(newVecSize);
    m_ioSteeringWheelAngleVec.resize(newVecSize);
    m_ioSteeringWheelTorqueVec.resize(newVecSize);
    m_ioSteeringAngleVec.resize(newVecSize);
    m_ioInverseSteeringRotVec.resize(newVecSize);
    m_ioThrottleVec.resize(newVecSize);
    m_ioBrakeVec.resize(newVecSize);
    m_ioBrakeTorqueActVec.resize(newVecSize);
    m_ioBrakeTorqueReqVec.resize(newVecSize);

    static int io_offset = 0;
    static float io_refresh_time = 0.0f;
    if (!m_showLive || io_refresh_time == 0.0f)
        io_refresh_time = ImGui::GetTime();

    // Update line plots
    while (io_refresh_time < ImGui::GetTime()) {
        // Vehicle IO Plots
        m_ioSpeedVec[io_offset]                 = test;
        m_ioSteeringWheelAngleVec[io_offset]    = test;
        m_ioSteeringWheelTorqueVec[io_offset]   = test;
        m_ioSteeringAngleVec[io_offset]         = test;
        m_ioInverseSteeringRotVec[io_offset]    = test;
        m_ioThrottleVec[io_offset]              = test;
        m_ioBrakeVec[io_offset]                 = test;
        m_ioBrakeTorqueActVec[io_offset]        = test;
        m_ioBrakeTorqueReqVec[io_offset]        = test;

        io_offset = (io_offset + 1) % static_cast<int>(newVecSize);
        io_refresh_time += 1 / m_windowRefreshRate;
    }

    ImGui::PlotLines("", m_ioSpeedVec.data(), newVecSize, io_offset, currentValue,
                             FLT_MAX, FLT_MAX, ImVec2(0, 150));
```
Basically, I am reducing the refresh rate by 1hz after every 10s increment in plot window length.

**Comment by ocornut** [comment]
Hello @salmanjalali, your issue is hard to understand without a self sufficient repro.

There is a 64K-vertices-per-mesh limit by default, which you can lift by redefining `ImDrawIdx` in imconfig.h, but reaching the limit normally should trigger an assert. Are your asserts enabled? 
Are other windows affected or just the window with the plot? What metrics do you see in `ShowMetricsWindow()` ?

Also the current Plot functions are designed to not render more segment/bar than the amount of pixel wide, so you would probably not reach the limit with only a few large plots.

**Comment by salmanjalali** [comment]
Thanks for the response.

I see, I am setting the line thickness to 3 pixels, which turns out to be why, and by setting it back to the default 1 pixel it does not happen.

Does this mean that I am trying to render more pixels than the screen size? Would you happen to know how I can override that?

My main intention is to make the plots more visible, as with a width of 1, on smaller screens it is almost impossible to see the plot.

Thanks

**Comment by ocornut** [comment]
> I am setting the line thickness to 3 pixels

Wow, what line thickness? Did you modify the `PlotLines()` code _and_ reported an issue with PlotLines without mentioning the fact that you modified the code?

You aren't answering the questions I asked, I can't move support further if you are ignoring my answer.

**Comment by salmanjalali** [comment]
That is my mistake, totally went over my head as I just changed one number in the code.

After adding the thickness the following is reported from the engine:
![after](https://user-images.githubusercontent.com/10891652/38513467-29bf47ee-3be3-11e8-8fe8-f4bb147a78b0.png)
As oppose to the default being:
![before](https://user-images.githubusercontent.com/10891652/38513473-2ec8ac76-3be3-11e8-927f-cda6e7611fb6.png)

**Comment by ocornut** [comment]
Thanks, that explains it. Your vertices count is REALLY high here. 

- You can raise the vertices limit by using 32-bit indices instead of 16-bits indice. See `ImDrawIdx` define in imconfig.h. That'll fix your issue but note that this is probably quite heavy to output every frame.

- When using thick lines with anti-aliasing enabled we need to create more rectangles per segment to provide AA on all edges, they are really heavier. Also the Plot function currently don't use a single path for the whole plot, which is an known issue (could be improved) so with thick lines we're wasting even more polygons on the four sides of each segments.
You may consider feeding less samples to the Plot functions (by making your vector smaller OR feeding a stride to the Plot function to skip over N samples every sample you display). 

- The plot functions can/should be improved in many ways to make this easier and lighter but right now there are very simple. I hope they'll get improved this year.

**Comment by salmanjalali** [comment]
Great, thank you so much for the info! I definitely am trying to render a lot. 

Not sure why the asserts are not triggering, but I will use one of your recommendations depending on what fits the application the most. (i.e. the looks and the load on the rest of the app)
